nav2_behaviors/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,167 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "assisted_teleop.hpp"
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_behaviors
+{
+AssistedTeleop::AssistedTeleop()
+: TimedBehavior<AssistedTeleopAction>(),
+  feedback_(std::make_shared<AssistedTeleopAction::Feedback>())
+{}
+
+void AssistedTeleop::onConfigure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // set up parameters
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "projection_time", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "input_vel_topic", rclcpp::ParameterValue(std::string("input_vel_topic")));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "joystick_topic", rclcpp::ParameterValue(std::string("joystick_topic")));
+
+  node->get_parameter("projection_time", projection_time_);
+  node->get_parameter("input_vel_topic", input_vel_topic_);
+  node->get_parameter("joystick_topic", joystick_topic_);
+}
+
+Status AssistedTeleop::onRun(const std::shared_ptr<const AssistedTeleopAction::Goal> command)
+{
+  command_time_allowance_ = command->time_allowance;
+  end_time_ = steady_clock_.now() + command_time_allowance_;
+
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  vel_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::inputVelocityCallback,
+      this, std::placeholders::_1));
+
+  joy_sub_ = node->create_subscription<sensor_msgs::msg::Joy>(
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::joyCallback,
+      this, std::placeholders::_1));
+
+  return Status::SUCCEEDED;
+}
+
+void AssistedTeleop::onActionCompletion()
+{
+  RCLCPP_INFO(logger_, "Action completed");
+  input_twist_ = geometry_msgs::msg::Twist();
+  vel_sub_.reset();
+  joy_sub_.reset();
+}
+
+Status AssistedTeleop::onCycleUpdate()
+{
+  rclcpp::Duration time_remaining = end_time_ - steady_clock_.now();
+  if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
+    stopRobot();
+    RCLCPP_WARN(
+      logger_,
+      "Exceeded time allowance before reaching the BackUp goal - Exiting BackUp");
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR(logger_, "Current robot pose is not available.");
+    return Status::FAILED;
+  }
+
+  // user states that teleop was successful
+  if (joy_.buttons.size() > 0 && joy_.buttons[0] > 0) {
```

Comment:
If joy is removed, all this can go too

Reply:
How will know if assisted teleop was completed successfully?

Reply:
The timeout or it is canceled, I suppose. Either way, we can't make any assumptions about joy commands, there's no standardization on what any particular button means or context, so this doesn't really work. Someone's node that will read in joystick commands and converts into ROS service / action / velocity commands will have to interpret and do the right thing.

Reply:
So in the case of a recovery if the teleoperator drives to a new location successfully he/she should be able to give the feedback that assisted teleop action was successful. Over wise, the behavior tree won't know if the assisted teleop action was successful. (Maybe we would always assume it completes successfully? If you can't teleop of the bot out pretty much out of options anyway...)

I was planning on adding parameters for the "success" and "failure" buttons that would be checked. So, it could be up to the user to make that decision.

Reply:
But, let me know what your thinking :)!

Reply:
> Maybe we would always assume it completes successfully?

I think that's a fair thing to do -- have it always return success in the BT node. A cancel just means to stop and isn't an indication of the successfulness (arguably, if cancelled, means it was successful and ready for navigation again)

Reply:
So, I thought a bit more about this. 

Even if we have the BT always return success we still don't have any feedback about when the action is done. Otherwise, we can only exit after a timeout. So I still think we need some way to allow the user to stop the action. 

I am not sure how we could implement this without adding another sub...